namer/rancher/src/main/scala/io/buoyant/namer/rancher/RancherNamer.scala

```diff
@@ -0,0 +1,56 @@
+package io.buoyant.namer.rancher
+
+import com.twitter.finagle.buoyant.ExistentialStability._
+import com.twitter.finagle._
+import com.twitter.finagle.stats.StatsReceiver
+import com.twitter.logging.Logger
+import com.twitter.util.Activity.State
+import com.twitter.util._
+
+class RancherNamer(
+  prefix: Path,
+  portMappings: Option[Map[String, Int]],
+  params: Stack.Params,
+  stats: StatsReceiver
+)(implicit val timer: Timer) extends Namer {
+  private[this] val log = Logger.get("io.buoyant.namer.rancher")
+
+  private[this] val ports = Map(
+    "http" -> 80,
+    "https" -> 443
+  ) ++ portMappings.getOrElse(Map())
+
+  private[this] val client = new RancherClient(log, params, stats)
+
+  private[this] def lookupPort(port: String): Option[Int] =
+    (Try(port.toInt).toOption, ports.get(port)) match {
+      case (Some(portNum), _) => Some(portNum)
+      case (_, Some(portNum)) => Some(portNum)
+      case _ => None
+    }
+
+  def lookup(path: Path): Activity[NameTree[Name]] = path.take(3) match {
+    case phd@Path.Utf8(port, stack, service) => lookupPort(port) match {
+      case Some(portNum) =>
+        val containers: Activity[Option[Addr]] = client.activity.map { allContainers =>
+          val eligable: Set[Address] = allContainers
+            .filter(c => stack.equalsIgnoreCase(c.stackName) && service.equalsIgnoreCase(c.serviceName))
+            .collect(scala.Function.unlift(_.toAddr(portNum)))
```

Comment:
I think you should be able to use flatMap here

Reply:
Clearly, I need to get better about fully understanding `flatMap`, because I think you are right.

Right now I use the `unlift`-method to turn `toAddr` into a partial function that only returns something when it returns `Some(address)`, and then `collect` to do a filter/map-combo which maps the partial function and filters out all the ones where it wasn't defined (i.e. the ones that would have had a `None` returned from `toAddr`.

I thought my way was fairly elegant – but I agree that `flatMap` aught to do the same and in a even shorter way. I just hadn't quite gotten that `Option` is traversable too, and returns either no or a single element, which can then be flattened into desired list too.